pkg/aspect/query/query.go

```diff
@@ -0,0 +1,170 @@
+*/
+
+package query
+
+import (
+	"aspect.build/cli/pkg/aspecterrors"
+	"aspect.build/cli/pkg/bazel"
+	"fmt"
+	"github.com/manifoldco/promptui"
+	"github.com/pkg/browser"
+	"github.com/spf13/cobra"
+	"io"
+	"io/ioutil"
+	"net/url"
+	"regexp"
+	"strings"
+
+	"aspect.build/cli/pkg/ioutils"
+)
+
+type PresetQuery struct {
+	Name        string
+	Description string
+	Query       string
+}
+
+type Query struct {
+	ioutils.Streams
+
+	bzl           bazel.Bazel
+	isInteractive bool
+
+	Presets   []*PresetQuery
+	ShowGraph bool
+}
+
+func New(streams ioutils.Streams, bzl bazel.Bazel, isInteractive bool) *Query {
+	return &Query{
+		Streams:       streams,
+		bzl:           bzl,
+		isInteractive: isInteractive,
+	}
+}
+
+func (q *Query) Run(_ *cobra.Command, args []string) error {
+	presets := make(map[string]*PresetQuery)
+	var names []string
+	for _, p := range q.Presets {
+		presets[p.Name] = p
+		names = append(names, fmt.Sprintf("%s: %s", p.Name, p.Description))
+	}
+
+	var preset *PresetQuery
+	if len(args) == 0 {
+		selectQueryPrompt := &promptui.Select{
+			Label: "Select a preset query",
+			Items: names,
+		}
+
+		i, _, err := selectQueryPrompt.Run()
+		if err != nil {
+			return fmt.Errorf("prompt failed: %w", err)
+		}
+
+		preset = q.Presets[i]
+	} else {
+		maybeQueryOrPreset := args[0]
+		if value, ok := presets[maybeQueryOrPreset]; ok {
+			// Treat this as the name of the preset query, so don't prompt for it.
+			fmt.Printf("Preset query \"%s\" selected\n", value.Name)
+			fmt.Printf("%s: %s\n", value.Name, value.Description)
+			preset = value
+		} else {
+			// Treat this as a raw query expression
+			return q.RunQuery(maybeQueryOrPreset, []string{})
+		}
+	}
+
+	if preset == nil {
+		return fmt.Errorf("unable to determine preset query")
+	}
+
+	query := preset.Query
+	placeholders := regexp.MustCompile(`(\?[a-zA-Z]*)`).FindAllString(query, -1)
+
+	if len(placeholders) > 0 {
+		if (len(placeholders) == len(args) - 1) {
+			for i, placeholder := range placeholders {
+				query = strings.ReplaceAll(query, placeholder, args[i+1])
+			}
+		} else {
+			for _, placeholder := range placeholders {
+				prompt := &promptui.Prompt{
+					Label: fmt.Sprintf("Value for '%s'", strings.TrimPrefix(placeholder, "?")),
+				}
+				val, err := prompt.Run()
+	
+				if err != nil {
+					return fmt.Errorf("prompt failed: %w", err)
+				}
+	
+				query = strings.ReplaceAll(query, placeholder, val)
+			}
+		}
+	}
+
+	return q.RunQuery(query, []string{})
+}
+
+func (q *Query) RunQuery(query string, flags []string) error {
+	if q.ShowGraph {
+		return q.RunQueryAndOpenResult(query, flags)
+	}
+
+	bazelCmd := []string{
+		"query",
+		query,
+	}
+
+	bazelCmd = append(bazelCmd, flags...)
+
+	if exitCode, err := q.bzl.Spawn(bazelCmd); exitCode != 0 {
+		err = &aspecterrors.ExitError{
+			Err:      err,
+			ExitCode: exitCode,
+		}
+		return err
+	}
+
+	return nil
+}
+
+func (q *Query) RunQueryAndOpenResult(query string, flags []string) error {
```

Comment:
I find this a strange way to divide code. I don't think you need `RunQueryAndOpenResult`. I think you only need `RunQuery`:

1. Check if `q.ShowGraph` and append `--output=graph`.
2. Execute Bazel like you execute here but check again for `q.ShowGraph` and based on the value you either do `io.Copy` from the pipe reader o the stdout (from the default streams) or consume in the way you are consuming now here.

If you want to separate the code for more intelligibility, you can then have 2 private functions that are specialized for printing to output or opening the graph viz.

Reply:
Function has been removed